tests/tests/block_tests.cpp

```diff
@@ -269,25 +276,36 @@ BOOST_AUTO_TEST_CASE( fork_blocks )
          BOOST_CHECK_EQUAL(db2.head_block_id().str(), b.id().str());
       }
 
-      //The two databases are on distinct forks now, but at the same height. Make a block on db2, make it invalid, then
+      //The two databases are on distinct forks now, but at the same height.
+      BOOST_CHECK_EQUAL(db1.head_block_num(), 13u);
+      BOOST_CHECK_EQUAL(db2.head_block_num(), 13u);
+      //Make a block on db2, make it invalid, then
       //pass it to db1 and assert that db1 doesn't switch to the new fork.
       signed_block good_block;
-      BOOST_CHECK_EQUAL(db1.head_block_num(), 13);
-      BOOST_CHECK_EQUAL(db2.head_block_num(), 13);
       {
          auto b = db2.generate_block(db2.get_slot_time(1), db2.get_scheduled_witness(1), init_account_priv_key, database::skip_nothing);
          good_block = b;
          b.transactions.emplace_back(signed_transaction());
          b.transactions.back().operations.emplace_back(transfer_operation());
          b.sign( init_account_priv_key );
-         BOOST_CHECK_EQUAL(b.block_num(), 14);
+         BOOST_CHECK_EQUAL(b.block_num(), 14u);
          GRAPHENE_CHECK_THROW(PUSH_BLOCK( db1, b ), fc::exception);
+         // Issue #938 make sure db is in a good state
+         fc::optional<signed_block> previous_block = db1.fetch_block_by_number(1);
+         BOOST_CHECK ( previous_block.valid() );
+         for( uint32_t curr_block_num = 2; curr_block_num <= 13; ++curr_block_num )
+         {
+            fc::optional<signed_block> curr_block = db1.fetch_block_by_number( curr_block_num );
+            BOOST_CHECK( curr_block.valid() );
+            BOOST_CHECK_EQUAL( curr_block->previous, previous_block->id() );
```

Comment:
This test should fail with the old code, but is not.

Reply:
`fetch_block_by_number` here will read data from `_fork_db`, thus will pass the check.

I think we can add more tests in the end: close db, reopen db, run this check again.

Reply:
Unfortunately, the original problem still does not show up. I haven't dug too deeply, but I'm guessing the reindexing fixes things when the database reopens.

Reply:
Oh, reopen at this point doesn't work. Try generate 30 more blocks to push the 2 blocks out of `fork_db`.

Reply:
Oh, seems due to empty blocks. I'm trying to update the test case.

Reply:
My comments earlier are incorrect.

Actually `PUSH_BLOCK( db1, good_block );` made all blocks stored in `db1` correct. I'm trying to reproduce.